src/pybind/mgr/deepsea/module.py

```diff
@@ -174,16 +182,27 @@ def describe_service(self, service_type, service_id, node_name):
 
         assert service_type in ("mon", "mgr", "mds", "rgw", None), service_type + " unsupported"
 
+        self.service_cache.remove_outdated()
+        if not self.service_cache.any_outdated() and not refresh:
+            # Let's hope the services are complete.
+            services = [orchestrator.ServiceDescription.from_json(d[1]) for d in self.service_cache.items_filtered([node_name])]
+            services = [s for s in services if
+                        (True if s.service_type is None else s.service_type == service_type) and
+                        (True if s.service_id is None else s.service_id == service_id)]
+            return orchestrator.TrivialReadCompletion(services)
+
+
         def process_result(event_data):
             result = []
             if event_data['success']:
                 for node_name, service_info in event_data["return"].items():
                     for service_type, service_instance in service_info.items():
-                        desc = orchestrator.ServiceDescription()
-                        desc.nodename = node_name
-                        desc.service_instance = service_instance
-                        desc.service_type = service_type
+                        desc = orchestrator.ServiceDescription(nodename=node_name,
+                                                               service_instance=service_instance,
+                                                               service_type=service_type)
                         result.append(desc)
+                for service in result:
+                    self.service_cache[service.nodename] = orchestrator.OutdatableData(service.to_json())
```

Comment:
This doesn't work properly when there is more than once service on a node, which there often will be, for example an mgr and mon on the same node. I'm not sure how best to rework this -- should the service cache just be a flat list? Or should it still be a dict keyed by node, but each item in the dict then becomes a list of services running on that node, rather than a single service?